designs/operation-input-output-and-unit-types.md

```diff
@@ -0,0 +1,373 @@
+
+1. Referencing the same input or output structure from multiple operations can
+   lead to backward-compatibility problems in the future if the inputs or
+   outputs of the operations ever need to diverge. By using the same structure,
+   teams unnecessarily tie the interfaces and future evolution of operations
+   together.
+2. Using the same structure for both input and output can lead to
+   backward-compatibility problems in the future if the members or traits used
+   in input needs to diverge from those used in output. Reuse between operations
+   is better facilitated by mixins, which allow members and traits to be shared,
+   but also allows teams to later refactor or even remove the applied mixins.
+3. Referencing an operation input or output shape as a member of another shape
+   gives the type multiple use cases. This makes it harder for code generators
+   to give special functionality to input and output types without affecting
+   the way nested types are handled, leading code generators to generate
+   synthetic types.
+
+**How will we get service teams to use the `@input` and `@output` traits?**
+
+* The inline operation input and output syntax being introduced in Smithy 2.0
+  is so convenient it will likely be used by most new models, and it will
+  automatically add these traits.
+* Model validation will emit a WARNING when operations target shapes that do
+  not have the `@input` or `@output` trait.
+
+**What happens to operations that have no input or output?**
+
+* Operations will by default target `smithy.api#Unit`.
+* A WARNING will be emitted when an operation defaults to `smithy.api#Unit`
+  without explicitly targeting `smithy.api#Unit`. This encourages modelers
+  to be deliberate in their commitment to never needing input or output in
+  the future for an operation.
+* Existing AWS service models will all be updated to define input or output if
+  they do not already.
+
+**Will this be a breaking change in the reference Java implementation?**
+
+No. The existing methods that return things like `Optional.empty` will continue
+to function exactly as they did before, but will be marked as deprecated. New
+methods will be added that always return a `ShapeId` or `StructureShape` when
+accessing things like the input of an operation or querying `OperationIndex`.
+
+**Won't it be annoying to have to model resource operations that reuse the same members?**
+
+No. Models typically SHOULD NOT use the same input or output structures for
+resources that perform CRUDL operations.
+
+1. The data returned in a Get operation SHOULD NOT be coupled to the data
+   returned in a List operation because getting the details of a resource
+   SHOULD require more elevated permissions than know a resource exists. If
+   List and Get responses are coupled in the model, sensitive members that
+   might need to be added to the output of a Get response will automatically
+   show up in the List response.
```

Comment:
A little of both.  It's quite common to use separate permission grants for listing vs getting resources as getting resources is meant to expose all the data of a resource.

I can expand the rationale to include your point too. If the argument is only about size and expense, then people think "I'll just reuse the structures because they're so small." But the problem is that you sometimes need to expose more sensitive information in order to get the resource than you'd want to expose in a list of the resources.